forte/data/extractor/base_extractor.py

```diff
@@ -0,0 +1,295 @@
+# limitations under the License.
+
+from abc import ABC
+import logging
+from typing import Tuple, Set, List, Dict, Any
+from typing import Union, Type, Hashable, Iterable, Optional
+from ft.onto.base_ontology import Annotation
+from forte.common.configuration import Config
+from forte.data.data_pack import DataPack
+from forte.data.vocabulary import Vocabulary
+from forte.data.converter.feature import Feature
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "BaseExtractor"
+]
+
+
+class BaseExtractor(ABC):
+    r"""The functionality of Extractor is as followed,
+            1. Build vocabulary.
+            2. Extract feature from datapack.
+            3. Remove feature in datapack.
+            3. Add prediction to datapack.
+
+        Explanation:
+            Vocabulary: Vocabulary is maintained as an inner class
+                in extractor. It will store the mapping from element
+                to index, which is an integer, and representation,
+                which could be an index integer or one-hot vector
+                depending on the configuration of the vocabulary.
+                Check :class:`forte.data.vocabulary.Vocabulary` for
+                more details.
+            Feature: A feature basically wraps the data we want from
+                one instance in a datapack. For example, the instance
+                can be one sentence in a datapack. Then the data wrapped
+                by the feature could be the token text of this sentence.
+                The data is already converted as list of indexes using
+                vocabulary. Besides the data, other information like the
+                raw data before indexing and some meta_data will also be
+                stored in the feature. Check
+                :class:`forte.data.converter.feature.Feature` for more
+                details.
+            Remove feature / Add prediction: Removing feature means remove
+                the exsiting data in the datapack. If we remove the feature
+                in the pack, then extracting feature will return empty list.
+                Adding prediction means we add the prediction from model
+                back to the datapack. If a datapack has some old data (for
+                example, the golden data in the test set), we can first
+                remove those data and then add our model prediction to
+                the pack.
+    Args:
+        config: An instance of `Dict` or
+            :class:`forte.common.configuration.Config` that provides all
+            configurable options. See :meth:`default_configs` for available
+            options and default values. Entry_type is the key that need to
+            be passed in and there will not be default value for this key.
+
+            entry_type: Type[Entry]. Required. The ontology type that the
+                extractor will get feature from.
+    """
+    VOCAB_ERROR_MSG = "When vocab_method is raw, vocabulary " \
```

Comment:
Can we make this as private member? (i.e. add prefix "_")

Reply:
Fixed.